src/core/ext/filters/client_channel/lb_policy.h

```diff
@@ -73,6 +73,16 @@ class LoadBalancingPolicy
     /// Closure to run when pick is complete, if not completed synchronously.
     /// If null, pick will fail if a result is not available synchronously.
     grpc_closure* on_complete;
+
+    // Callback set by lb policy to be notified of trailing metadata.
+    // The callback must be scheduled on grpc_schedule_on_exec_ctx.
+    grpc_closure* recv_trailing_metadata_ready;
```

Comment:
I'm not sure what's the best way to handle ownership of the metadata.  At first I thought this closure should be initialized to run on `grpc_combiner_scheduler(args_.combiner)`, but then I notice that by the time it runs the metadata has already been freed. Should I instead make a copy of the metadata, or bump the refs of the existing metadata, and have the closure be responsible for freeing it?

Reply:
Do we expect that in most cases `recv_trailing_metadata_ready` will inspect the metadata without making any changes to the LB policy? If that is the case, would we want to avoid chaining every RPC on the combiner scheduler?

To be specific, I'm thinking is `recv_trailing_metadata_ready` can run on `grpc_schedule_on_exec_ctx` but the LB is not synchronized. If the trailer causes us to read/write the LB, then it is the closure's responsibility for scheduling a separate closure on `grpc_combiner_scheduler(args_.combiner)`. Does this make sense or is it a premature optimization?

Reply:
The two use-cases we know of for the LB policy accessing trailing metadata are:

1. The xds LB policy wants to see the trailing metadata to know when the call is complete, which it needs to record stats to be sent back to the balancer.

2. There may be LB policies that use this information for routing affinity.  For example, we could initially send a request to some random backend, which could forward the request to the right backend and then send back trailing metadata telling the client that all future requests for that resource should be sent to a different backend.  This would require the LB policy to update its routing affinity table for use in subsequent requests.

Although there may be other uses in the future, I would guess that pretty much any time the LB policy wants to know about the trailing metadata, it's going to use that information to modify its internal state, which will require synchronization.  So I think we should handle this for it rather than making the LB policy implementor handle it.

BTW, I just realized that case 2 is also interesting in that the LB policy will want to remove the metadata from the batch before letting the batch proceed back to the application.  This is another reason why we need to make sure the LB policy's callback has finished before we pass the batch back up to the surface.